🌱 Annotate nodes with cluster info

CecileRobertMichon opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**: Adds basic cluster info to nodes as annotations to allow users to know if the node was created using cluster api and which cluster/namespace/machine or machine pool it belongs to.

Tested using release-0.3 branch:

```
 k --kubeconfig ./kubeconfig describe node capz-cluster-7-md-0-5f9fq
Name:               capz-cluster-7-md-0-5f9fq
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=Standard_D2s_v3
                    beta.kubernetes.io/os=linux
                    failure-domain.beta.kubernetes.io/region=southcentralus
                    failure-domain.beta.kubernetes.io/zone=0
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=capz-cluster-7-md-0-5f9fq
                    kubernetes.io/os=linux
                    node.kubernetes.io/instance-type=Standard_D2s_v3
                    topology.kubernetes.io/region=southcentralus
                    topology.kubernetes.io/zone=0
Annotations:        cluster.x-k8s.io/cluster-name: capz-cluster-7
                    cluster.x-k8s.io/cluster-namespace: default
                    cluster.x-k8s.io/machine: capz-cluster-7-md-0-567986c6b8-xk8k8
                    cluster.x-k8s.io/owner-kind: MachineSet
                    cluster.x-k8s.io/owner-name: capz-cluster-7-md-0-567986c6b8
                    kubeadm.alpha.kubernetes.io/cri-socket: /run/containerd/containerd.sock
                    node.alpha.kubernetes.io/ttl: 0
                    projectcalico.org/IPv4Address: 10.1.0.9/16
                    projectcalico.org/IPv4VXLANTunnelAddr: 192.168.194.128
                    volumes.kubernetes.io/controller-managed-attach-detach: true
``` 

```
 k --kubeconfig ./kubeconfig describe node capz-cluster-b-mp-0000001
Name:               capz-cluster-b-mp-0000001
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=Standard_D2s_v3
                    beta.kubernetes.io/os=linux
                    failure-domain.beta.kubernetes.io/region=southcentralus
                    failure-domain.beta.kubernetes.io/zone=1
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=capz-cluster-b-mp-0000001
                    kubernetes.io/os=linux
                    node.kubernetes.io/instance-type=Standard_D2s_v3
                    topology.kubernetes.io/region=southcentralus
                    topology.kubernetes.io/zone=1
Annotations:        cluster.x-k8s.io/cluster-name: capz-cluster-b
                    cluster.x-k8s.io/cluster-namespace: default
                    cluster.x-k8s.io/owner-kind: MachinePool
                    cluster.x-k8s.io/owner-name: capz-cluster-b-mp-0
                    kubeadm.alpha.kubernetes.io/cri-socket: /run/containerd/containerd.sock
                    node.alpha.kubernetes.io/ttl: 0
                    projectcalico.org/IPv4Address: 10.1.0.7/16
                    projectcalico.org/IPv4VXLANTunnelAddr: 192.168.219.0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
```                 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4044

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -85,6 +86,14 @@ func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv
+	annotations.SetNodeAnnotation(node, clusterv1.ClusterNameAnnotation, machine.Spec.ClusterName)
+	annotations.SetNodeAnnotation(node, clusterv1.ClusterNamespaceAnnotation, machine.GetNamespace())
+	annotations.SetNodeAnnotation(node, clusterv1.MachineAnnotation, machine.Name)
+	if err := remoteClient.Patch(ctx, node, client.Merge); err != nil {
+		log.V(2).Info("Failed patch node to set annotations", "err", err, "node name", node.Name)
+	}
```
If we're not using our patch helper, we should probably check if there are any changes here before performing the patch, otherwise we'd end up patching at every reconciliation

CecileRobertMichon replied:
done

CecileRobertMichon commented:
/retest

MHC flake

vincepri commented on util/annotations/helpers.go:
```diff
@@ -49,3 +50,11 @@ func HasWithPrefix(prefix string, annotations map[string]string) bool {
 }
+
+// SetNodeAnnotation sets a key value annotation on the Node.
+func SetNodeAnnotation(node *corev1.Node, key, value string) {
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+	node.Annotations[key] = value
+}
```
Could this be on a metav1 object instead?

CecileRobertMichon replied:
indeed, changed

vincepri commented on exp/controllers/machinepool_controller_noderef.go:
```diff
@@ -100,6 +101,21 @@ func (r *MachinePoolReconciler) reconcileNodeRefs(ctx context.Context, cluster *
+		annotations.SetNodeAnnotation(node, clusterv1.ClusterNameAnnotation, mp.Spec.ClusterName)
+		annotations.SetNodeAnnotation(node, clusterv1.ClusterNamespaceAnnotation, mp.GetNamespace())
+		annotations.SetNodeAnnotation(node, clusterv1.MachinePoolAnnotation, mp.Name)
+		if err := clusterClient.Patch(ctx, node, client.Merge); err != nil {
```
Same comment as above, re: patch helper

detiber commented on controllers/machine_controller_noderef.go:
```diff
@@ -85,6 +86,14 @@ func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv
+	// Reconcile node annotations.
+	annotations.SetNodeAnnotation(node, clusterv1.ClusterNameAnnotation, machine.Spec.ClusterName)
+	annotations.SetNodeAnnotation(node, clusterv1.ClusterNamespaceAnnotation, machine.GetNamespace())
+	annotations.SetNodeAnnotation(node, clusterv1.MachineAnnotation, machine.Name)
```
Would it also be helpful to annotate the Node with the kind/name of the owner if a controller reference exists on the Machine here?

That would help link the Node more directly to a MachineSet or KubeadmControlPlane (if the Machine is managed by either of those).

CecileRobertMichon replied:
Added

CecileRobertMichon commented:
Thanks for the review @vincepri @detiber. The change ended up being a bit bigger than I expected because I added some logic to fetch the Machine owner but I think it's better now. Added unit test coverage for the new helper/utils functions.

CecileRobertMichon commented on util/util.go:
```diff
@@ -335,6 +341,54 @@ func GetOwnerMachine(ctx context.Context, c client.Client, obj metav1.ObjectMeta
+func GetMachineOwner(ctx context.Context, c client.Client, obj metav1.Object) (*MachineOwner, error) {
+	allowedGKs := []schema.GroupKind{
+		{
+			Group: clusterv1.GroupVersion.Group,
```
/hold

I need to fix this to use the controlplane Group. Unfortunately, there's an import cycle between controlplane/api and util. I will fix the cycle in a separate PR to unblock this.

detiber replied:
I'm not sure there is a need to enforce certain allowed GVKs here, any controller reference should be sufficient for our purpose, since it would indicate ownership of the lifecycle of the resource (that it wasn't created by a user).

I do wonder if instead of just adding the Kind as the value of the annotation if it would make more sense to use GVK instead to avoid any potential for name conflicts/confusion.

CecileRobertMichon replied:
removed this func altogether

CecileRobertMichon replied:
/hold cancel

detiber commented on exp/controllers/machinepool_controller_noderef.go:
```diff
@@ -100,6 +101,25 @@ func (r *MachinePoolReconciler) reconcileNodeRefs(ctx context.Context, cluster *
+			clusterv1.MachinePoolAnnotation:      mp.Name,
+		}
+		if annotations.AddAnnotations(node, desired) {
+			if err := clusterClient.Patch(ctx, node, client.Merge); err != nil {
```
Similar to above wrt `MergeFrom`

detiber commented on controllers/machine_controller_noderef.go:
```diff
@@ -85,6 +86,22 @@ func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv
+		clusterv1.ClusterNamespaceAnnotation: machine.GetNamespace(),
+		clusterv1.MachineAnnotation:          machine.Name,
+	}
+	if owner, err := util.GetMachineOwner(ctx, r.Client, machine); err == nil && owner != nil {
+		desired[clusterv1.OwnerKindAnnotation] = owner.GetKind()
+		desired[clusterv1.OwnerNameAnnotation] = owner.GetName()
+	}
```
I don't think we actually need to go this far, I think we could do something like this instead:

```suggestion
	if owner := metav1.GetControllerOfNoCopy(obj); owner != nil {
		desired[clusterv1.OwnerKindAnnotation] = owner.Kind
		desired[clusterv1.OwnerNameAnnotation] = owner.Name
	}
```

CecileRobertMichon replied:
True, this works and is a lot simpler 😅 thanks!

fabriziopandini commented on api/v1alpha4/common_types.go:
```diff
@@ -31,6 +31,24 @@ const (
+	MachineAnnotation = "cluster.x-k8s.io/machine"
+
+	// MachinePoolAnnotation is the annotation set on nodes identifying the machine pool the node belongs to.
+	MachinePoolAnnotation = "cluster.x-k8s.io/machine-pool"
```
I think that in case of MachinePools we should not add a different annotation but instead add something like

```
cluster.x-k8s.io/owner-kind: MachinePool
cluster.x-k8s.io/owner-name: capz-cluster-2-mp-0
```

So all the nodes managed by higher-level abstraction have a consistent set of labels.
The only exception for the machine pools is the `cluster.x-k8s.io/machine` being missing

vincepri replied:
Given that these are annotations, could we instead json-marshal the owner reference as we see it today instead of creating a new standard to store this information?

CecileRobertMichon replied:
what do you mean by json-marshal the owner reference? How would it be represented in the annotations? Would it be one annotation for the entire owner ref?

vincepri replied:
Something like `cluster.x-k8s.io/owner` and its value would be the json value of an OwnerReference object

CecileRobertMichon replied:
IMO 

```
cluster.x-k8s.io/owner-kind: MachinePool
cluster.x-k8s.io/owner-name: capz-cluster-2-mp-0
``` 

is more readable than dumping the json string in there. I don't really care either way, what I really wanted from this PR in the first place was the cluster name, owner name + kind were just added on for additional info.

vincepri replied:
Sounds good to me, I assume that would depend on who's using these and how, for example if this is only for human users that's probably fine. For machines / controllers the information above for example is missing the API group (this is only a problem if we ever say that a Machine could be controlled outside of Cluster API main types)

vincepri commented on controllers/machine_controller_noderef.go:
```diff
@@ -85,6 +86,22 @@ func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv
+		desired[clusterv1.OwnerNameAnnotation] = owner.GetName()
+	}
+	if annotations.AddAnnotations(node, desired) {
+		if err := remoteClient.Patch(ctx, node, client.Merge); err != nil {
```
Could we potentially also make use of the patch helper here?

CecileRobertMichon replied:
@vincepri in this case what value does patch helper bring that can't be accomplished by MergeFrom?

vincepri replied:
Patch helper checks if there are any changes before issuing the patch, IIRC merge from won't unless you explicitly check

CecileRobertMichon replied:
switched to patch helper

vincepri reviewed (approved):
/lgtm
/milestone v0.4.0

fabriziopandini commented:
/lgtm

detiber commented:
/approve
